flash/core/model.py

```diff
@@ -188,3 +210,111 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         return []
+
+    ### THE FOLLOWING IS A POC FOR DISTRIBUTED PREDICTION
+    def on_predict_start(self):
+        # TODO: Add hook to lightning Trainer
+        if self.data_pipeline is not None:
+            self.data_pipeline._attach_to_model(self)
+
+        if self.postprocessing_pipeline is not None:
+            self.postprocessing_pipeline._attach_to_model(self)
+
+    def predict_step(self, batch, batch_idx):
+        # TODO: Move lightning predict loop from predict to predict_step
+        if isinstance(batch, (tuple, list)) and len(batch) == 2:
+            x, y = batch
+        else:
+            x, y = batch, None
+
+        return self(x)
+
+    def new_predict(
+        self,
+        x: Any,
+        skip_collate: Optional[bool] = None,
+        data_pipeline: Optional[DataPipeline] = None,
+        postprocessing_pipeline: Optional[PostProcessingPipeline] = None,
+        data_loader_kwargs: Optional[dict] = None,
+        **trainer_kwargs
+    ):
+        if data_pipeline is not None:
+            self.data_pipeline = data_pipeline
+        if postprocessing_pipeline is not None:
+            self.postprocessing_pipeline = postprocessing_pipeline
+
+        trainer = self._create_trainer('predict', **trainer_kwargs)
+
+        if data_loader_kwargs is None:
+            data_loader_kwargs = {}
+
+        if 'num_workers' not in data_loader_kwargs:
+            # leave one for main process
+            data_loader_kwargs['num_workers'] = os.cpu_count() - 1
+
+        auto_collate = None
+        if 'collate_fn' not in data_loader_kwargs:
+            auto_collate = not skip_collate
+
+        dl = self.data_pipeline._generate_loader(x, auto_collate=auto_collate, **data_loader_kwargs)
+
+        return trainer.predict(self, dl)
+
+    def _create_trainer(self, stage: str, **trainer_kwargs):
+        # TODO: Also use these for trainer creation in training?
+        # TODO: Have default trainer kwargs per task?
+        _trainer_kwargs = {}
+        # TODO: Adjust this to trainer running stage from pl
+        if stage == 'predict':
+            _trainer_kwargs.update(logger=None)
+
+        if not 'gpus' in trainer_kwargs and not 'tpu_cores' in trainer_kwargs:
+            _trainer_kwargs['gpus'], _trainer_kwargs['tpu_cores'] = self._parse_default_devices()
+
+        _trainer_kwargs.update(trainer_kwargs)
+
+        if not hasattr(self, 'trainer') or self.trainer is None or self._last_trainer_kwargs != trainer_kwargs:
+            self._last_trainer_kwargs = _trainer_kwargs
+            self.trainer = None
+            return Trainer(**_trainer_kwargs)
+
+        else:
+            return self.trainer
+
+    def _parse_default_devices(self):
+        gpus = None,
+        tpu_cores = None
+
+        if torch.cuda.is_available():
+            gpus = torch.cuda.device_count()
+
+        # TODO: Add logic for automatted TPU device parsing
+
+        return gpus, tpu_cores
+
+    def serve(
```

Comment:
I think this serve function is confusing. Serve means different things on the context and should not be used to perform prediction. It is more like a mode.

Reply:
Yes I thought we wanted to reserve serve for actually setting up a hosting server for the model.

Reply:
okay, so how do we want to handle it? I mean technically the user can set those defaults manually as well, but I feel like this is not what flash should be aiming for.